File: python/open_guardrail/guards/token_smuggling.py
"""Detect token smuggling via alternate encodings."""

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_DEFAULT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(
        r"\batob\s*\(", re.IGNORECASE,
    ),
    re.compile(
        r"Buffer\.from\s*\(", re.IGNORECASE,
    ),
    re.compile(
        r"base64[._\-]?decode", re.IGNORECASE,
    ),
    re.compile(
        r"\\x[0-9a-fA-F]{2}(\\x[0-9a-fA-F]{2}){3,}",
    ),
    re.compile(
        r"(?:0x[0-9a-fA-F]{2}\s*,?\s*){4,}",
    ),
    re.compile(
        r"\brot13\b", re.IGNORECASE,
    ),
    re.compile(
        r"\\u[0-9a-fA-F]{4}(\\u[0-9a-fA-F]{4}){3,}",
    ),
    re.compile(
        r"\bu(?:\+|%)[0-9a-fA-F]{4}", re.IGNORECASE,
    ),
    re.compile(
        r"1gn0r3|byp[a@]ss|3x3cut3|d3l3t3",
        re.IGNORECASE,
    ),
    re.compile(
        r"[\u0400-\u04ff].*[a-zA-Z]|[a-zA-Z].*[\u0400-\u04ff]",
    ),
    re.compile(
        r"[\uff01-\uff5e]{3,}",
    ),
]


class _TokenSmuggling:
    def __init__(
        self,
        *,
        action: str = "block",
        extra_patterns: Optional[List[re.Pattern[str]]] = None,
    ) -> None:
        self.name = "token-smuggling"
        self.action = action
        self._patterns = list(_DEFAULT_PATTERNS)
        if extra_patterns:
            self._patterns.extend(extra_patterns)

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in self._patterns:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = min(len(matched) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="token-smuggling",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Token smuggling detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "matched_patterns": len(matched),
                    "reason": (
                        "Text contains alternate encoding"
                        " patterns that may smuggle"
                        " hidden instructions"
                    ),
                }
                if triggered
                else None
            ),
        )


def token_smuggling(
    *,
    action: str = "block",
    extra_patterns: Optional[List[re.Pattern[str]]] = None,
) -> _TokenSmuggling:
    return _TokenSmuggling(
        action=action, extra_patterns=extra_patterns
    )

File: packages/guards/src/token-smuggling.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface TokenSmugglingOptions {
  action: 'block' | 'warn';
}

const DEFAULT_PATTERNS: RegExp[] = [
  /atob\s*\(/i,
  /Buffer\.from\s*\(/i,
  /base64[\s\S]{0,30}decode/i,
  /decode[\s\S]{0,30}base64/i,
  /\\x[0-9a-f]{2}(\\x[0-9a-f]{2}){3,}/i,
  /0x[0-9a-f]{2}[\s,]+0x[0-9a-f]{2}[\s,]+0x[0-9a-f]{2}/i,
  /rot13/i,
  /\\u00[0-9a-f]{2}(\\u00[0-9a-f]{2}){3,}/i,
  /[\.\-]{3,}\s*[\.\-]{3,}\s*[\.\-]{3,}/,
  /1[gG][nN]0[rR]3|byp[a4][s5]{2}|r[uU]1[eE][s5]/i,
  /[\u0400-\u04ff][\u0000-\u007f]|[\u0000-\u007f][\u0400-\u04ff]/,
  /[\uff21-\uff3a\uff41-\uff5a]{4,}/,
];

export function tokenSmuggling(options: TokenSmugglingOptions): Guard {
  return {
    name: 'token-smuggling',
    version: '0.1.0',
    description: 'Detects token smuggling attacks via alternate encodings',
    category: 'security',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of DEFAULT_PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'token-smuggling',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}
